File: flow-client/src/main/java/com/vaadin/client/flow/collection/JsArray.java
/*
 * Copyright 2000-2026 Vaadin Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.vaadin.client.flow.collection;

import jsinterop.annotations.JsOverlay;
import jsinterop.annotations.JsPackage;
import jsinterop.annotations.JsProperty;
import jsinterop.annotations.JsType;

import com.vaadin.client.flow.collection.JsCollections.ForEachCallback;

/**
 * Native JS Array binding. Use {@link JsCollections#array()} to create an
 * instance.
 *
 * @author Vaadin Ltd
 * @since 1.0
 * @param <T>
 *            the type of the array items
 */
@JsType(isNative = true, name = "Array", namespace = JsPackage.GLOBAL)
public class JsArray<T> {

    /**
     * Should not be directly created.
     */
    protected JsArray() {
        // prevent direct instantiation
    }

    /**
     * Gets the item at the given index. This is corresponding to
     * <code>return array[index]</code> in JavaScript.
     */
    @JsOverlay
    public final T get(int index) {
        return JsniHelper.getValueNative(this, index);
    }

    /**
     * Sets the item at the given index. This is corresponding to
     * <code>array[index] = value</code> in JavaScript.
     */
    @JsOverlay
    public final void set(int index, T value) {
        JsniHelper.setValueNative(this, index, value);
    }

    /**
     * Adds items to the end of this array.
     *
     * @return the new length of the array
     */
    public native int push(@SuppressWarnings("unchecked") T... values);

    /**
     * Adds items to the end of this array.
     *
     * @return the new length of the array
     */
    @JsOverlay
    public final int pushArray(JsArray<? extends T> values) {
        return JsniHelper.pushArray(this, values);
    }

    /**
     * Gets the current length of this array.
     */
    @JsProperty(name = "length")
    public native int length();

    /**
     * Removes and adds a number of items at the given index.
     */
    @JsOverlay
    public final JsArray<T> spliceArray(int index, int remove,
            JsArray<? extends T> add) {
        return JsniHelper.spliceArray(this, index, remove, add);
    }

    /**
     * Removes and adds a number of items at the given index.
     */
    public native JsArray<T> splice(int index, int remove,
            @SuppressWarnings("unchecked") T... add);

    /**
     * Removes the item at the given index.
     */
    @JsOverlay
    @SuppressWarnings("unchecked")
    public final T remove(int index) {
        return splice(index, 1).get(0);
    }

    /**
     * Clears the array.
     */
    @JsOverlay
    public final JsArray<T> clear() {
        JsniHelper.clear(this);
        return this;
    }

    /**
     * Checks if the array is empty (length == 0).
     */
    @JsOverlay
    public final boolean isEmpty() {
        return length() == 0;
    }

    /**
     * Removes the given item from the array.
     *
     * @return <code>true</code> if the item was found and removed from the
     *         array, <code>false</code> otherwise
     */
    @JsOverlay
    public final boolean remove(T toRemove) {
        for (int i = 0; i < length(); i++) {
            if (get(i) == toRemove) {
                remove(i);
                return true;
            }
        }
        return false;
    }

    /**
     * Removes and returns the first value from the array.
     */
    public native T shift();

    /**
     * Invokes the provided callback for each value in this array.
     */
    public native void forEach(ForEachCallback<T> callback);

}

// Helper for stuff not allowed in a @JsType class
class JsniHelper {
    private JsniHelper() {
        // Only static stuff here, should never be instantiated
    }

    static native <T> int pushArray(JsArray<T> array,
            JsArray<? extends T> values)
    /*-{
        return array.push.apply(array, values);
    }-*/;

    static native <T> JsArray<T> spliceArray(JsArray<T> array, int index,
            int remove, JsArray<? extends T> add)
    /*-{
        return array.splice.apply(array, [index, remove].concat(add));
    }-*/;

    static native void clear(JsArray<?> array)
    /*-{
        array.length = 0;
    }-*/;

    static native <T> T getValueNative(JsArray<T> array, int index)
    /*-{
        return array[index];
    }-*/;

    static native <T> void setValueNative(JsArray<T> array, int i, T value)
    /*-{
        array[i] = value;
    }-*/;

}

File: flow-client/src/main/frontend/internal/client/flow/collection/JsArray.ts
/*
 * Copyright 2000-2026 Vaadin Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

/**
 * Native JS array operations ported from the `JsArray<T>` interface of
 * JsArray.java. In the GWT client `JsArray<T>` was a thin wrapper over a native
 * JavaScript array (with a separate JRE implementation for server-side use); in
 * TypeScript a plain array already is that native array, so no wrapper is needed
 * and only the operations that are more than a single native call are provided
 * here as free functions.
 *
 * The full public Java API maps as follows:
 *
 * ```
 * Native TypeScript array syntax (intentionally NOT wrapped here):
 *   get(index)                    -> array[index]
 *   set(index, value)             -> array[index] = value
 *   push(...values)               -> array.push(...values)
 *   length()                      -> array.length
 *   splice(index, remove, ...add) -> array.splice(index, remove, ...add)
 *   shift()                       -> array.shift()
 *   forEach(callback)             -> array.forEach(callback)
 *
 * Exported free-function helpers (below), because they are genuinely more than a
 * native operation or bundle several native calls:
 *   pushArray, spliceArray, clear, isEmpty, remove, removeItem
 * ```
 *
 * The private `JsniHelper` statics behind the Java `@JsOverlay` methods are not
 * ported.
 *
 * Java's `JsArray<T>` carried a `@param <T> the type of the array items`. That
 * type parameter is preserved on the helpers whose item argument or return type
 * depends on it (`pushArray`, `spliceArray`, `clear`, `remove`, `removeItem`);
 * it is omitted from `isEmpty`, where the element type appears only once in a
 * single argument and a generic would add nothing.
 */

/**
 * Appends every value (spread) onto the array, returning the new length.
 * Mirrors the public JsArray.pushArray overlay.
 *
 * @typeParam T - the type of the array items
 * @param array - the array to operate on
 * @param values - the new values to add
 * @returns the new length of the array
 */
export function pushArray<T>(array: T[], values: readonly T[]): number {
  return array.push(...values);
}

/**
 * Splices the array at index, removing `remove` elements and inserting the
 * `add` values (spread), returning the removed elements. Mirrors the public
 * JsArray.spliceArray overlay.
 *
 * @typeParam T - the type of the array items
 * @param array - the array to operate on
 * @param index - the index at which do do the operation
 * @param remove - the number of items to remove
 * @param add - new items to add
 * @returns an array of removed items
 */
export function spliceArray<T>(array: T[], index: number, remove: number, add: readonly T[]): T[] {
  return array.splice(index, remove, ...add);
}

/**
 * Removes the item at the given index and returns it. Mirrors the public
 * JsArray.remove(int index) overlay (which is `splice(index, 1).get(0)`).
 *
 * @typeParam T - the type of the array items
 * @param array - the array to operate on
 * @param index - the index to remove
 * @returns the remove item
 */
export function remove<T>(array: T[], index: number): T {
  return array.splice(index, 1)[0];
}

/**
 * Empties the array and returns it, mirroring the public JsArray.clear overlay
 * (which returns the cleared array).
 *
 * @typeParam T - the type of the array items
 * @param array - the array to operate on
 * @returns the cleared array
 */
export function clear<T>(array: T[]): T[] {
  array.length = 0;
  return array;
}

/**
 * Checks if the array is empty (length === 0). Mirrors the public
 * JsArray.isEmpty overlay.
 *
 * @param array - the array to operate on
 * @returns true if the array is empty, false otherwise
 */
export function isEmpty(array: unknown[]): boolean {
  return array.length === 0;
}

/**
 * Removes the first item that is identical (`===`, matching Java reference
 * `==`) to the given value, returning whether one was found and removed. Maps
 * to the by-value Java overload JsArray.remove(T toRemove); it is named
 * `removeItem` here because TypeScript cannot overload `remove` on a single
 * argument whose type may itself be a number.
 *
 * @typeParam T - the type of the array items
 * @param array - the array to operate on
 * @param toRemove - the item to remove
 * @returns `true` if the item was found and removed from the array, `false` otherwise
 */
export function removeItem<T>(array: T[], toRemove: T): boolean {
  for (let i = 0; i < array.length; i++) {
    if (array[i] === toRemove) {
      array.splice(i, 1);
      return true;
    }
  }
  return false;
}
